circuit/types/integers/src/add_checked.rs

```diff
@@ -73,37 +73,38 @@ impl<E: Environment, I: IntegerType> AddChecked<Self> for Integer<E, I> {
             }
-        } else {
-            // TODO: Optimize
+        } else if I::is_signed() {
             // Instead of adding the bits of `self` and `other` directly, the integers are
             // converted into a field elements, and summed, before converting back to integers.
             // Note: This is safe as the field is larger than the maximum integer type supported.
             let sum = self.to_field() + other.to_field();
 
-            // Extract the integer bits from the field element, with a carry bit.
-            let (sum, carry) = match sum.to_lower_bits_le(I::BITS as usize + 1).split_last() {
-                Some((carry, bits_le)) => (Integer::from_bits_le(bits_le), carry.clone()),
+            // Extract the integer bits from the field element, ignoring the carry bit as it is not relevant for signed addition.
+            let sum = match sum.to_lower_bits_le(I::BITS as usize + 1).split_last() {
+                Some((_, bits_le)) => Integer::from_bits_le(bits_le),
                 // Note: `E::halt` should never be invoked as `I::BITS as usize + 1` is greater than zero.
                 None => E::halt("Malformed sum detected during integer addition"),
             };
 
-            // Check for overflow.
-            match I::is_signed() {
-                // For signed addition, overflow and underflow conditions are:
-                //   - a > 0 && b > 0 && a + b < 0 (Overflow)
-                //   - a < 0 && b < 0 && a + b > 0 (Underflow)
-                //   - Note: if sign(a) != sign(b) then over/underflow is impossible.
-                //   - Note: the result of an overflow and underflow must be negative and positive, respectively.
-                true => {
-                    let is_same_sign = self.msb().is_equal(other.msb());
-                    let is_overflow = is_same_sign & sum.msb().is_not_equal(self.msb());
-                    E::assert_eq(is_overflow, E::zero());
-                }
-                // For unsigned addition, ensure the carry bit is zero.
-                false => E::assert_eq(carry, E::zero()),
-            }
+            // For signed addition, overflow and underflow conditions are:
+            //   - a > 0 && b > 0 && a + b < 0 (Overflow)
+            //   - a < 0 && b < 0 && a + b > 0 (Underflow)
+            //   - Note: if sign(a) != sign(b) then over/underflow is impossible.
+            //   - Note: the result of an overflow and underflow must be negative and positive, respectively.
+            let is_same_sign = self.msb().is_equal(other.msb());
+            let is_overflow = is_same_sign & sum.msb().is_not_equal(self.msb());
+            E::assert_eq(is_overflow, E::zero());
+
+            sum
+        } else {
+            // Instead of adding the bits of `self` and `other` directly, witness the integer sum.
+            let sum: Integer<E, I> = witness!(|self, other| self.add_wrapped(&other));
```

Comment:
I think that `add_wrapped` generates an extra R1CS boolean variable for the carry. Since we do checked addition here, we can save that variable by doing something like what's done above for the signed case, allocating just `I::BITS` variables for the result.

Reply:
This is inside the `witness!` macro, so only `I::BITS` variables should get allocated no?

Reply:
Yes, I missed that it was done inside the `witness!`.